### PR TITLE
dynamo/torchxla integration: trace on xla rather than eager

### PR DESCRIPTION
**Pull request by shunting314**
In #87741 we added the inference support for dynamo/torchxla integration. Later on in #88449 we attempt to add the training support. That attempt is not smooth because
- we try 2 things together
   1. let dynamo trace the model on xla rather than eager
   2. enable training
- It turns out neither of these two tasks are trivial enough.

Furthermore, item 2 (enable training) depends on item 1 (tracing on xla). We enable training via AOTAutograd. AOTAutograd lift all model parameters/buffers as graph inputs. Without item 1 being done, we would need copy all graph inputs (including model parameters/buffers) from eager device to xla devices. That hurts performance a lot. Have a cache to map eager parameter to XLA parameter does not solve the problem since the update on either will not sync automatically to the other. They will easily go out of sync.

This PR let dynamo trace the model on XLA rather than eager. This is a preparation step to enabling training.

Also, tracing on XLA makes the data movement more efficient. We see 1.5x geomean speedup compared to previous 1.38x.
```
+-------------------------+--------------------+-------------------------+
| Model                   |   XLA (trace once) |   XLA (trace everytime) |
+=========================+====================+=========================+
| resnet18                |            1.38    |                 1.008   |
+-------------------------+--------------------+-------------------------+
| resnet50                |            1.227   |                 0.998   |
+-------------------------+--------------------+-------------------------+
| resnext50_32x4d         |            1.544   |                 1.008   |
+-------------------------+--------------------+-------------------------+
| alexnet                 |            1.085   |                 1.045   |
+-------------------------+--------------------+-------------------------+
| mobilenet_v2            |            2.028   |                 1.013   |
+-------------------------+--------------------+-------------------------+
| mnasnet1_0              |            1.516   |                 0.995   |
+-------------------------+--------------------+-------------------------+
| squeezenet1_1           |            0.868   |                 1.01    |
+-------------------------+--------------------+-------------------------+
| vgg16                   |            1.099   |                 1.008   |
+-------------------------+--------------------+-------------------------+
| BERT_pytorch            |            3.26    |                 1.027   |
+-------------------------+--------------------+-------------------------+
| timm_vision_transformer |            2.182   |                 1.015   |
+-------------------------+--------------------+-------------------------+
| geomean                 |            1.50389 |                 1.01261 |
+-------------------------+--------------------+-------------------------+
```

Example command
```
GPU_NUM_DEVICES=1 python benchmarks/dynamo/torchbench.py --randomize-input --performance --trace-on-xla --only resnet18 --backend=torchxla_trace_once
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [inline torch/_dynamo/optimizations/torchxla_integration.py:22]: I'll cleanup these debugging code.

File: benchmarks/dynamo/common.py
Comment (shunting314) [inline benchmarks/dynamo/common.py:1872]: When debugging some tricky issue like https://github.com/pytorch/xla/issues/4174 , I found it's quite helpful to let torchbench run a simple enough model (rather than a model from torchbench).

This piece of code is not the idea UX. I'm thinking something like:
```
--only=path:<path to model file>;model:<ModelClass>
```
So torchbench will load the model from the path and call ModelClass.get_example_inputs to retrieve the inputs.

Would like to know other's opinion. @jansel  @wconstab

File: benchmarks/dynamo/common.py
Comment (wconstab) [inline benchmarks/dynamo/common.py:412]: i think we should be doing the mark step on the input boundary of the dynamo-xla backend, in order to ensure our graph-tracing starts on fresh xlatensors not xlatensors that have already accumulated some trace.  I think you're also working on that change in parallel.  Can we just get that working first and then not need this?

File: benchmarks/dynamo/common.py
Comment (wconstab) [reply benchmarks/dynamo/common.py:1872]: that sounds like a useful feature to me. I also use my own ToyModel classes a lot for various bringup work.  I think the way you proposed (with --only) is pretty good.  The hardcoded "linear" thing is probably not worth landing.

File: benchmarks/dynamo/common.py
Comment (shunting314) [reply benchmarks/dynamo/common.py:412]: @wconstab hmm, I think we still need it.

E.g., if we skip this mark_step call and do it inside the optimized function, then that optimized function will incur overhead of computing the randomized inputs. The other call we compare with does not have this overhead. 

Then the comparison would be unfair.

Comment (shunting314) [reply benchmarks/dynamo/common.py:412]: We add `mark_step` or sync call in a few other places for the baseline. Since we trace on xla, the baseline is also running on XLA, we need explicitly sync tensors for it.

Comment (shunting314) [reply benchmarks/dynamo/common.py:412]: @wconstab  I added a sync when we enter the optimized function.

But I think we may not need a mandate sync when we leave the function. Leave it as it is have the advantage that we potentially accumulate a large graph. Let me know if there are other concerns.

Comment (wconstab) [reply benchmarks/dynamo/common.py:412]: ahh.. when we exit the optimized function, isn't the last thing we do to invoke an API that executes the cached xla graph?  that API may be equivalent to mark_step in all the ways that matter.  all of the tensors we return from the optimized function should be 
 - XLATensors
 - backed by devicedata, not IR value
 
 so any operations performed in eager land after one of our optimized functions returns should be starting fresh graphs.  If you returned XLAtensor backed by IR value from the optimized function, i don't think the xla runtime would behave the way you expect if you trace additional IR onto the graph and then execute it.

at least for the first design, i think it is best to break the XLA graph on entry/exit of the dynamo function

Comment (shunting314) [reply benchmarks/dynamo/common.py:412]: yep, for now the output of the optimized function will most likely be device data. Adding a mark_step should not matter much since it's like a nop. I can add one to be safe.

Comment (wconstab) [reply benchmarks/dynamo/common.py:412]: ok. i don't actually want to have an extra one. it might incur some small pointless overhead. it's probably safe to have it.  it's worth reviewing this part of the design with @JackCaoG and make sure we're not missing anything. but if we can prove the tensors we're returning from the execute-cached-graph api are fresh in all cases, i think we don't need the output mark step.

Comment (shunting314) [reply benchmarks/dynamo/common.py:412]: This extra `mark_step` to sync the inputs is needed only for benchmarking. If we don't want it, we need make sure both runs (baseline and optimized) incur the same overhead of executing the graph that computes the inputs.

From other scenario (e.g. prod case if there is any in future), people don't need add such a call for sure.

File: benchmarks/dynamo/common.py
Comment (JackCaoG) [inline benchmarks/dynamo/common.py:251]: :D if model is on XLA_DEVICE, it is almost guarantee that computation won't be executed by just running the model(unless all ops fallback), I guess we can make a stronger statement here.

File: benchmarks/dynamo/common.py
Comment (JackCaoG) [reply benchmarks/dynamo/common.py:412]: If we run the cached computation directly, no additional IR should be generated since we skip the tracing part.
If we run the model on lazy for the first time, we already call `xla_sync_multi` to sync all results of the models. The only IR that might not be materialized should be the IR seed.

I think we can use `torch_xla._XLAC. _xla_tensors_report(0, str(xla_device))` before the `mark_step` to confirm my theory. If there are additional pending graphs I felt like it might be an issue and we should address it.

Comment (wconstab) [reply benchmarks/dynamo/common.py:412]: > This extra mark_step to sync the inputs is needed only for benchmarking. If we don't want it, we need make sure both runs (baseline and optimized) incur the same overhead of executing the graph that computes the inputs.

ok. for benchmarking reasons, you may be right. is mark_step synchronous though? i thought mark_step was just going to break the graph and schedule the execution.  you'd also need to ensure the execution finished.  also, i think there is a point in the dynamo benchmark where cuda.sync happens before/after measurements for this same reason. maybe that is where you should move this maybe_mark_step thing?

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (JackCaoG) [inline torch/_dynamo/optimizations/torchxla_integration.py:106]: Would be interesting to use `torch_xla._XLAC. _xla_tensors_report(0, str(xla_device))` to figure out what's the pending graph here. I think for the case that dynamo can not capture the whole graph, the pending graph is those model that fallback to lazy device. If user manually runs parts of the model on xla directly, this will also collapse those execution which I think is the correct behavior we want.

File: benchmarks/dynamo/common.py
Comment (JackCaoG) [reply benchmarks/dynamo/common.py:412]: `mark_step` is not synchronous by defualt, there a arg you can toggle the `wait` behavior. You are right that we actually need to wait for previous computation to finish. In lazy code we achieved that by trying to grab the deviceContext's lock. We are currently not doing that, so let me draft a quick pr to do something similar to https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1035 in our `_run_cached_graph` api.

Comment (JackCaoG) [reply benchmarks/dynamo/common.py:412]: Instead of making `mark_step` synchronous here, I think it is better to make `_run_cached_graph` grab the device lock. This way we can overlap the dynamo bridges that prepare inputs to `_run_cached_graph` and the device execution.

Comment (shunting314) [reply benchmarks/dynamo/common.py:412]: > also, i think there is a point in the dynamo benchmark where cuda.sync happens before/after measurements for this same reason.

yep, I think I know which of the code you mean. But one difference here is, for the cuda.sync, we want its execution time to be included to the benchmark execution time. But for this `mark_step`, we want to exclude it.

**Comment by shunting314** [comment]
I may address comments on this PR a bit slow since I want to make sure training works based on this PR first. Sorry for potential delays.

File: benchmarks/dynamo/common.py
Comment (JackCaoG) [reply benchmarks/dynamo/common.py:412]: now I think about it a bit more there are two cases

1. When we first time tracing the graph. This one is actually because we trace using the lazy and lock thing should already been taking care of.
2. When we run the cached graph, since we already run `torch_xla._XLAC._xla_sync_multi(args, [])`, we know that at least that input to the cached graph is materalized. However in a very rare chance that there are still some pending computations on XLA device. We can not execute more than 1 graph at any given time, so `_run_cahce_graph` should still try to grab the device lock before proceed to execution.

Comment (JackCaoG) [reply benchmarks/dynamo/common.py:412]: https://github.com/pytorch/xla/pull/4195

File: benchmarks/dynamo/common.py
Comment (shunting314) [reply benchmarks/dynamo/common.py:252]: hmm, but what if we run a cached graph? In that case, the result will be DeviceData nodes and the computation is already performed?

Comment (JackCaoG) [reply benchmarks/dynamo/common.py:252]: cached graph is a bit ambigous. If it is a cache graph in lazy concept, we still do the tracing of the IR graph, but only skip the `IR->HLO->executable`. In this case, without a `mark_step`, result will still be a pending IR graph.

If we are talking about cached graph in dynamo concept, we actually skip all tracing and execute the graph directly. I thought the comment "running the model" means under lazy concept.

Comment (shunting314) [reply benchmarks/dynamo/common.py:252]: This method is used for both baseline and optimized code path to measure execution time. So we should expect both scenarios you mentioned

Comment (JackCaoG) [reply benchmarks/dynamo/common.py:252]: sg!

File: benchmarks/dynamo/common.py
Comment (shunting314) [reply benchmarks/dynamo/common.py:1873]: Here is the separate PR for this: https://github.com/pytorch/pytorch/pull/89028 . @wconstab

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration.py:106]: @JackCaoG  The symptom is the model inputs does not show up as xla graph inputs. On the other hand I actually don't see pending graph from the API you mentioned.

Here is the command for you reference (repro with the Linear model):
```
GPU_NUM_DEVICES=1 python benchmarks/dynamo/torchbench.py --randomize-input --performance --trace-on-xla --backend=torchxla_trace_once --only path:/pytorch/myscripts/model_collection.py,class:LinearModel
```

Output:
```
Tensor: id=1, shape=f32[10,10]{1,0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10,10]{1,0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=2, shape=f32[10]{0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10]{0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=3, shape=f32[10,10]{1,0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10,10]{1,0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=4, shape=f32[10]{0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10]{0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=5, shape=f32[2,10]{1,0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[2,10]{1,0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=6, shape=f32[10,10]{1,0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10,10]{1,0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=7, shape=f32[10]{0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10]{0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=8, shape=f32[10,10]{1,0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10,10]{1,0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=9, shape=f32[10]{0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[10]{0} xla::device_data(), device=GPU:0, ROOT=0
}


Tensor: id=10, shape=f32[2,10]{1,0}, device=GPU:0, ir_nodes=1
IR {
  %0 = f32[2,10]{1,0} xla::device_data(), device=GPU:0, ROOT=0
}
```

Comment (JackCaoG) [reply torch/_dynamo/optimizations/torchxla_integration.py:106]: hmm OK, doesn't seem like we have any pending computation. I think we can ignore this for now since `mark_step` is a good workaround. We can revisit later.

File: benchmarks/dynamo/common.py
Comment (wconstab) [inline benchmarks/dynamo/common.py:289]: the check is loose (any) - should it be tight? (all)

Comment (shunting314) [reply benchmarks/dynamo/common.py:289]: yep, but any valid case should not mix xla/eager tensors. So for valid cases, checking 'all' or 'any' should be equivalent. I could add assertion to guard against invalid (mix tensor type) cases. Let me know if you want me to do that.

File: benchmarks/dynamo/common.py
Comment (wconstab) [inline benchmarks/dynamo/common.py:283]: why don't we also sync_multi here if XLA? i think it is going to work fine as currently written, since it is unlikely there is an xla computation running before the first `timed` call.  but it breaks with tradition, the sync here is 'defensive'

File: benchmarks/dynamo/common.py
Comment (wconstab) [inline benchmarks/dynamo/common.py:1413]: are you landing this separately?

Comment (shunting314) [reply benchmarks/dynamo/common.py:1413]: yep, this is supposed to be landed by #89028 . But that PR is blocked by some 'Mac OS' CI issue which affects multiple PRs.

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (wconstab) [inline torch/_dynamo/optimizations/torchxla_integration.py:163]: update this comment?  you don't need to convert xla to eager anymore.  but you still have a timer for inplace copy_ for syncing any input mutations back to input tensors right?

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (wconstab) [inline torch/_dynamo/optimizations/torchxla_integration.py:153]: better to rename copy_args_ts to ... idk... inplace_update_ts ?

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (wconstab) [inline torch/_dynamo/optimizations/torchxla_integration.py:136]: I prefer you rename `ivalue` to `xla_value` or `device_data` (something to avoid confusing people, since this is _not_ using torchscript

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration.py:163]: Good catch.  The timer for 'copy result' should be removed together with the comments. For 'inplace updates' yes, we still have a timer for that.

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration.py:153]: yep, will do

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (wconstab) [inline torch/_dynamo/optimizations/torchxla_integration.py:75]: is this a good place to do an assert that all xla_args and xla_model are on xla device?  Did you do that somewhere else (i may have missed it if it's not highlighted as a diff here)

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration.py:153]: Or args_inplace_update_ts to emphasis that we are doing this for arguments :) ?

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration.py:136]: I'll call it xla_value then.

File: torch/_dynamo/optimizations/torchxla_integration.py
Comment (shunting314) [reply torch/_dynamo/optimizations/torchxla_integration.py:75]: I believe there is no redundant checks in codepath reach here. We can add that.

**Review by wconstab — APPROVED** [review]
Minor comments, but you're fixing them. stamp to unblock. thanks! LGTM

**Review by JackCaoG — APPROVED** [review]
Thanks! Working on the xla side test fix.

**Comment by JackCaoG** [comment]
Hey Shunting, can you update https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt to `e9434c7504ceae4d6c317936a97932c426a582d9` which is the commit hash for https://github.com/pytorch/xla/pull/4205/commits . More details in https://github.com/pytorch/xla/tree/master/.circleci#coodinating-merges-for-breaking-pytorch-prs

**Review by jansel — APPROVED** [review]

**Comment by shunting314** [comment]
Working with Jack on #4205 to resolve the CI failure for some xla tests. Will commit once we resolve the test failures.

**Pull request closed without merging**
